quickwit-indexing/src/actors/packager.rs

```diff
@@ -205,17 +219,47 @@ fn create_packaged_split(
         .sum();
 
     // Extracts tag values from `_tags` special fields.
-    let mut tags = BTreeSet::default();
     let index_reader = split
         .index
         .reader_builder()
         .reload_policy(ReloadPolicy::Manual)
         .try_into()?;
+    let mut tag_fields_to_token_count: HashMap<String, usize> = HashMap::default();
+    for reader in index_reader.searcher().segment_readers() {
+        for (tag_field_name, tag_field) in tag_fields_list {
+            let inv_index = reader.inverted_index(*tag_field)?;
+            let token_count = tag_fields_to_token_count
+                .entry(tag_field_name.clone())
+                .or_insert(0);
+            *token_count += inv_index.terms().num_terms() as usize;
+        }
+    }
+    ctx.record_progress();
+
+    let mut tags = BTreeSet::default();
     for reader in index_reader.searcher().segment_readers() {
         let inv_index = reader.inverted_index(tags_field)?;
         let mut terms_streamer = inv_index.terms().stream()?;
         while let Some((term_data, _)) = terms_streamer.next() {
-            tags.insert(String::from_utf8_lossy(term_data).to_string());
+            let tag_value = String::from_utf8_lossy(term_data).to_string();
+            if let Some(field_name) = extract_field_name_from_tag_value(&tag_value)
+                .and_then(|field_name| {
+                    tag_fields_to_token_count
+                        .get(&field_name)
+                        .map(|token_count| (field_name, token_count))
+                })
+                .and_then(|(field_name, token_count)| {
+                    if token_count >= &MAX_VALUES_PER_TAG_FIELD {
+                        Some(field_name)
+                    } else {
+                        None
+                    }
+                })
+            {
+                tags.insert(make_too_many_tag_value(&field_name));
+            } else {
+                tags.insert(tag_value);
+            }
```

Comment:
this part is annoying. Why not iterating of the field tokens directly? This may sound a little bit stupid because we need to rebuild the tag but it would be more readable.

Reply:
At the end, I would do something like that:
```rust
for (tag_field_name, &tag_field) in tag_fields_list {
  if segments.iter().map(|seg| seg.inv_index(tag_field).terms().num_terms()).sum()> 1000 {
      continue;
  }
  // code to add tokens
}

```